packages/foam-core/src/model/note.ts

```diff
@@ -30,8 +31,20 @@ export interface NoteLinkDefinition {
   position?: Position;
 }
 
-export interface Note {
+export interface BaseResource {
   uri: URI;
+}
+
+export interface Attachment extends BaseResource {
+  type: 'attachment';
+}
+
+export interface Placeholder extends BaseResource {
+  type: 'placeholder';
+}
+
+export interface Note extends BaseResource {
+  type: 'note';
```

Comment:
Suggestion here, using an enum for the different types. Something like:

```ts
enum ResourceType {
  Attachement = 'ATTACHEMENT',
  Placeholder = 'PLACEHOLDER',
  Note = 'NOTE',
}
```

IMHO it makes it easier to reference and to define behaviors depending on the type, if needed. For example:

```ts
function renderResource(resource: Resource): void {
  switch (resource.type) {
    case ResourceType.Attachement:
      renderAttachement(resource); // resource is already type Attachement
      break;
    case ResourceType.Placeholder:
      renderPlaceholder(resource);
      break;
    case ResourceType.Note:
      renderNote(resource);
  }
}
```

Reply:
Adding a detail, why doesn't `BaseResource` have a type which then gets overriden?

Reply:
because that's the pattern to do type unions in typescript: https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions

Having the field on `BaseResource` as string would allow any value there. having it as a union of the 3 strings is no different than what I have written, but making changes would require the two sets to be consistent.

The reason why `BaseResource` exists is because `uri` is everywhere, so instead of adding it 3 times I put it in the base class, but I'd rather move `uri` to the 3 resource types than move type to `BaseResource` - do you know what I mean?

Regarding using the enum, I am not opposed to it per-se, it just feels heavier than a simple union type, and I am not sure it has advantages (you can do the switch with the union type as well, and TS also detects the `never` type).
That being said, I also see VSCode uses that a lot (although I don't love it), and am open to other opinions.

Just came across this, which describes pretty well my preference for union types: https://blog.bam.tech/developer-news/should-you-use-enums-or-union-types-in-typescript

Reply:
Some valid points in this article indeed. I guess both are fine. Just wondering - are you planning on exporting those types and using them from `foam-vscode` as well? In this case enums might make it clearer on the possible type values, as you'd have to import it as well :thinking: But maybe that's just the pattern I'm used to.

Reply:
yes probably that will be part of the Foam API that will be available to other extensions built on top of us.
what do you mean by enum being easier? asking because at times in vscode I actually find them hard to deal with, but also in my case I feel it's just a matter of habit

Reply:
I meant enum are easier to import/export, so you don't have to retype the type in another package and can rely solely on the enum instead.